File: Homeworks/hw2_java/src/main/java/com/pascal/hw2_java/Calculator.java
package com.pascal.hw2_java;

public class Calculator {
    public final static String NONE = "none";

    private double result = 0;
    private String currentNum = "0";
    private String operation = NONE;

    public String clearAll() {
        result = 0;
        currentNum = "0";
        operation = NONE;
        return "0";
    }

    public String removeDigit() {
        if (currentNum.length() == 1) {
            currentNum = "0";
        } else {
            currentNum = currentNum.substring(0, currentNum.length() - 1);
        }
        return currentNum;
    }

    public String addDotSymbol() {
        if (!currentNum.contains(".") && currentNum.length() != 0) {
            currentNum += ".";
        }
        return currentNum;
    }

    public String calcResult() {
        if (!isDouble(currentNum)) return currentNum;
        if (NONE.equals(operation)) return currentNum;
        double firstOperand = result;
        double secondOperand = Double.parseDouble(currentNum);
        double calculusResult = 0;
        try {
            switch (operation) {
                case "+": {
                    calculusResult = firstOperand + secondOperand;
                    break;
                }
                case "-": {
                    calculusResult = firstOperand - secondOperand;
                    break;
                }
                case "*": {
                    calculusResult = firstOperand * secondOperand;
                    break;
                }
                case "/": {
                    calculusResult = firstOperand / secondOperand;
                    break;
                }
                case "^": {
                    calculusResult = Math.pow(firstOperand, secondOperand);
                    break;
                }
            }
        } catch (ArithmeticException e) {
            e.printStackTrace();
            calculusResult = Double.NaN;
        }
        operation = NONE;
        result = calculusResult;
        currentNum = String.valueOf(calculusResult);
        return currentNum;
    }

    public String getCurrentNum() {
        return currentNum;
    }

    public void setOperation(String oper) {
        if (!isDouble(currentNum)) return;

        if (NONE.equals(operation)) {
            result = Double.parseDouble(currentNum);
            currentNum = "0";
        }
        operation = oper;
    }

    public String addDigit(String digit) {
        if ("0".equals(currentNum))
            currentNum = "";
        currentNum += digit;
        return currentNum;
    }

    public boolean isDouble(String string) {
        if ("NaN".equals(currentNum)) return false;
        try {
            double num = Double.parseDouble(string);
        } catch (Exception e) {
            return false;
        }
        return true;
    }
}

File: Homeworks/hw2_kotlin/src/main/java/com/pascal/hw2_kotlin/Calculator.kt
package com.pascal.hw2_kotlin

import kotlin.math.pow

class Calculator {
    var result = 0.0
    var currentNum = "0"
    private var operation = "none"

    fun clearAll(): String {
        result = 0.0
        currentNum = "0"
        operation = "none"
        return "0"
    }

    fun removeDigit(): String {
        currentNum = if (currentNum.length == 1) "0" else currentNum.substring(0, currentNum.length - 1)
        return currentNum
    }

    fun addDotSymbol(): String {
        if (!currentNum.contains(".") && currentNum.isNotEmpty()) {
            currentNum += "."
        }
        return currentNum
    }

    fun calcResult(): String {
        if (!isParseble(currentNum)) return currentNum
        if ("none" == operation) return currentNum
        val firstOperand = result
        val secondOperand = currentNum.toDouble()
        var calculusResult = 0.0
        try {
            when (operation) {
                "+" -> calculusResult = firstOperand + secondOperand
                "-" -> calculusResult = firstOperand - secondOperand
                "*" -> calculusResult = firstOperand * secondOperand
                "/" -> calculusResult = firstOperand / secondOperand
                "^" -> calculusResult = firstOperand.pow(secondOperand)
            }
        } catch (e: ArithmeticException) {
            e.printStackTrace()
            calculusResult = Double.NaN
        }
        operation = "none"
        result = calculusResult
        currentNum = calculusResult.toString() + ""
        return currentNum
    }

    fun getOperation(): String {
        return operation
    }

    fun setOperation(oper: String) {
        if (!isParseble(currentNum)) return
        if ("none" == operation) {
            result = currentNum.toDouble()
            currentNum = "0"
        }
        operation = oper
    }

    fun addDigit(digit: String): String {
        if ("0" == currentNum) currentNum = ""
        currentNum += digit
        return currentNum
    }

    private fun isParseble(string: String): Boolean {
        if ("NaN" == currentNum) return false
        try {
            val num = string.toDouble()
        } catch (e: Exception) {
            return false
        }
        return true
    }
}